### PR TITLE
509: Update API expected body from String to OB data model object

### PR DESCRIPTION
**Pull request by andra-racovita**
Issue: https://github.com/SecureBankingAccessToolkit/SecureBankingAccessToolkit/issues/509

- Added differentiated controllers for each API and for each version.
- Changed response and request object to use the appropriate open banking objects for each version.
- Fixed unit tests for the new API response and request objects.

### CODE REVIEW

File: securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticpayments/CalculateResponseElements.java [hunk 1]
```diff
@@ -44,17 +45,14 @@ public interface CalculateResponseElements {
     @RequestMapping(value = "/calculate-elements",
             produces = {"application/json; charset=utf-8"},
             method = RequestMethod.POST)
-    ResponseEntity calculateElements(
+    OBWriteDomesticConsentResponse5 calculateElements(
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticpayments/CalculateResponseElements.java:48]: Needs to be ResponseEntity<T> `Extension of HttpEntity that adds a HttpStatus status code. Used in RestTemplate as well @Controller methods.`
- ResponseEntity<OBWriteDomesticConsentResponse5>

File: securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticscheduledpayments/CalculateResponseElements.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            @ApiResponse(code = 405, message = "Method Not Allowed"),
+            @ApiResponse(code = 406, message = "Not Acceptable"),
+            @ApiResponse(code = 429, message = "Too Many Requests"),
+            @ApiResponse(code = 500, message = "Internal Server Error", response = OBErrorResponse1.class)
+    })
+    @RequestMapping(value = "/calculate-elements",
+            produces = {"application/json; charset=utf-8"},
+            method = RequestMethod.POST)
+    OBWriteDomesticScheduledConsentResponse5 calculateElements(
+            @ApiParam(value = "Create an Account Access Consents", required = true)
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticscheduledpayments/CalculateResponseElements.java:49]: Needs to be ResponseEntity<T> `Extension of HttpEntity that adds a HttpStatus status code. Used in RestTemplate as well @Controller methods.`
- ResponseEntity<OBWriteDomesticScheduledConsentResponse5>

File: securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticstandingorders/CalculateResponseElements.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            @ApiResponse(code = 404, message = "Not found"),
+            @ApiResponse(code = 404, message = "Method Not Allowed"),
+            @ApiResponse(code = 406, message = "Not Acceptable"),
+            @ApiResponse(code = 429, message = "Too Many Requests"),
+            @ApiResponse(code = 400, message = "Internal Server Error", response = OBErrorResponse1.class)
+    })
+    @RequestMapping(value = "/calculate-elements",
+            produces = {"application/json; charset=utf-8"},
+            method = RequestMethod.POST)
+    OBWriteDomesticStandingOrderConsentResponse6 calculateElements(
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticstandingorders/CalculateResponseElements.java:48]: Needs to be ResponseEntity<T> `Extension of HttpEntity that adds a HttpStatus status code. Used in RestTemplate as well @Controller methods.`
- ResponseEntity<OBWriteDomesticStandingOrderConsentResponse6>

File: securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/internationalpayments/CalculateResponseElements.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            @ApiResponse(code = 404, message = "Method Not Allowed"),
+            @ApiResponse(code = 406, message = "Not Acceptable"),
+            @ApiResponse(code = 429, message = "Too Many Requests"),
+            @ApiResponse(code = 400, message = "Internal Server Error", response = OBErrorResponse1.class)
+    })
+    @RequestMapping(value = "/calculate-elements",
+            produces = {"application/json; charset=utf-8"},
+            method = RequestMethod.POST)
+    OBWriteInternationalConsentResponse6 calculateElements(
+            @ApiParam(value = "Create an Account Access Consents", required = true)
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/internationalpayments/CalculateResponseElements.java:49]: Needs to be ResponseEntity<T> `Extension of HttpEntity that adds a HttpStatus status code. Used in RestTemplate as well @Controller methods.`
- ResponseEntity<OBWriteInternationalConsentResponse6>

**Review by jorgesanchezperez — CHANGES_REQUESTED** [review]
Update all signature of calculateElements operation method to use **ResponseEntity<T>**:  `Extension of HttpEntity that adds a HttpStatus status code. Used in RestTemplate as well @Controller methods.`
- Example ResponseEntity<OBWriteInternationalConsentResponse6> calculateElements(..)

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/v3_1_4/domesticpayments/CalculateResponseElementsController.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            OBWriteDomesticConsent4 body,
+            String intent,
+            String xFapiFinancialId,
+            String xFapiAuthDate,
+            String xFapiCustomerIpAddress,
+            String xFapiInteractionId,
+            HttpServletRequest request) throws OBErrorResponseException {
+        try {
+            return (OBWriteDomesticConsentResponse4) PaymentConsentGeneral.calculate(body, intent, xFapiFinancialId, request);
+        } catch (UnsupportedOperationException | JsonProcessingException e) {
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/v3_1_4/domesticpayments/CalculateResponseElementsController.java:53]: When update the calculateElements signature to use ResponseEntity<T> the return needs to be similar to below to encapsulate the http status.
return ResponseEntity.status(HttpStatus.OK).body(
     PaymentConsentGeneral.calculate(body, intent, xFapiFinancialId, request)
);

**Review by jorgesanchezperez — CHANGES_REQUESTED** [review]
When update the calculateElements signature to use `ResponseEntity<T>` the return needs to be similar to below to add the `HttpStatus` in the response.
**Example**
```java
return ResponseEntity.status(HttpStatus.OK).body(
     PaymentConsentGeneral.calculate(body, intent, xFapiFinancialId, request)
);
```

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/v3_1_4/domesticpayments/CalculateResponseElementsController.java [hunk 6]
```diff
@@ -49,7 +50,7 @@ public OBWriteDomesticConsentResponse4 calculateElements(
             String xFapiInteractionId,
             HttpServletRequest request) throws OBErrorResponseException {
         try {
-            return (OBWriteDomesticConsentResponse4) PaymentConsentGeneral.calculate(body, intent, xFapiFinancialId, request);
+            return ResponseEntity.status(HttpStatus.OK).body((OBWriteDomesticConsentResponse4)PaymentConsentGeneral.calculate(body, intent, xFapiFinancialId, request));
```
Comment (jorgesanchezperez) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/v3_1_4/domesticpayments/CalculateResponseElementsController.java:53]: Please format the response line to be more readable.

File: securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticpayments/CalculateResponseElements.java (hunk 1 above)
Comment (mariantiris) [reply securebanking-openbanking-uk-rs-backoffice-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/calculate/elements/v3_1_5/domesticpayments/CalculateResponseElements.java]: Changes implemented

**Review by jorgesanchezperez — CHANGES_REQUESTED** [review]
Please, format the return lines to be more readable.

**Review by jorgesanchezperez — APPROVED** [review]

**Pull request merged**